src/signTxOutput.c

```diff
@@ -0,0 +1,651 @@
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+
+		signTx_handleOutput_address_ui_runStep();
+	}
+}
+
+enum {
+	HANDLE_OUTPUT_ADDRESS_PARAMS_STEP_DISPLAY_SPENDING_PATH = 350,
+	HANDLE_OUTPUT_ADDRESS_PARAMS_STEP_DISPLAY_STAKING_INFO,
+	HANDLE_OUTPUT_ADDRESS_PARAMS_STEP_DISPLAY_AMOUNT,
+	HANDLE_OUTPUT_ADDRESS_PARAMS_STEP_RESPOND,
+	HANDLE_OUTPUT_ADDRESS_PARAMS_STEP_INVALID,
+};
+
+static void signTx_handleOutput_addressParams_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	ui_callback_fn_t* this_fn = signTx_handleOutput_addressParams_ui_runStep;
+
+	ASSERT(subctx->stateData.output.outputType == OUTPUT_TYPE_ADDRESS_PARAMS);
+
+	UI_STEP_BEGIN(subctx->ui_step);
+
+	UI_STEP(HANDLE_OUTPUT_ADDRESS_PARAMS_STEP_DISPLAY_SPENDING_PATH) {
+		ui_displayPathScreen("Send to address", &subctx->stateData.output.params.spendingKeyPath, this_fn);
+	}
+	UI_STEP(HANDLE_OUTPUT_ADDRESS_PARAMS_STEP_DISPLAY_STAKING_INFO) {
+		ui_displayStakingInfoScreen(&subctx->stateData.output.params, this_fn);
+	}
+	UI_STEP(HANDLE_OUTPUT_ADDRESS_PARAMS_STEP_DISPLAY_AMOUNT) {
+		ui_displayAdaAmountScreen("Send", subctx->stateData.output.adaAmount, this_fn);
+	}
+	UI_STEP(HANDLE_OUTPUT_ADDRESS_PARAMS_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+
+		advanceState();
+	}
+	UI_STEP_END(HANDLE_OUTPUT_ADDRESS_PARAMS_STEP_INVALID);
+}
+
+static void signTx_handleOutput_addressParams()
+{
+	ASSERT(subctx->stateData.output.outputType == OUTPUT_TYPE_ADDRESS_PARAMS);
+
+	security_policy_t policy = policyForSignTxOutputAddressParams(
+	                                   commonTxData->isSigningPoolRegistrationAsOwner,
+	                                   &subctx->stateData.output.params,
+	                                   commonTxData->networkId, commonTxData->protocolMagic
+	                           );
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
+
+	{
+		// add to tx
+		uint8_t addressBuffer[MAX_ADDRESS_SIZE];
+		size_t addressSize;
+		addressSize = deriveAddress(
+		                      &subctx->stateData.output.params,
+		                      addressBuffer,
+		                      SIZEOF(addressBuffer)
+		              );
+		ASSERT(addressSize > 0);
+		ASSERT(addressSize < BUFFER_SIZE_PARANOIA);
+
+		txHashBuilder_addOutput_basicData(
+		        txHashBuilder,
+		        addressBuffer, addressSize,
+		        subctx->stateData.output.adaAmount,
+		        subctx->numAssetGroups
+		);
+	}
+
+	{
+		// select UI steps
+		switch (policy) {
+#	define  CASE(POLICY, UI_STEP) case POLICY: {subctx->ui_step=UI_STEP; break;}
+			CASE(POLICY_SHOW_BEFORE_RESPONSE, HANDLE_OUTPUT_ADDRESS_PARAMS_STEP_DISPLAY_AMOUNT);
+			CASE(POLICY_ALLOW_WITHOUT_PROMPT, HANDLE_OUTPUT_ADDRESS_PARAMS_STEP_RESPOND);
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+
+		signTx_handleOutput_addressParams_ui_runStep();
+	}
+}
+
+static void signTxOutput_handleBasicDataAPDU(uint8_t* wireDataBuffer, size_t wireDataSize)
```

Comment:
"basic" is kind of an umbrella term, IMHO a name like `signTxOutput_handleTopLevelDataAPDU` would do a better job conveying the essence of what's going on here